File: scripts/mcts_evaluate.py
import argparse
import pathlib
import random

from utttpy.game.ultimate_tic_tac_toe import UltimateTicTacToe
from utttpy.selfplay.monte_carlo_tree_search import (
    MonteCarloTreeSearch,
    serialize_evaluated_state,
    serialize_evaluated_actions,
)


def run_argparse() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument("--uttt_state", type=str, required=True)
    parser.add_argument("--num_simulations", type=int, required=True)
    parser.add_argument("--exploration_strength", type=float, required=True)
    parser.add_argument("--random_seed", type=int, required=True)
    parser.add_argument("--output_path", type=pathlib.Path, required=True)
    args = parser.parse_args()
    return args


def main() -> None:
    args = run_argparse()
    print(args)

    random.seed(args.random_seed)

    uttt = UltimateTicTacToe(state=bytearray(map(int, args.uttt_state)))

    mcts = MonteCarloTreeSearch(
        uttt=uttt,
        num_simulations=args.num_simulations,
        exploration_strength=args.exploration_strength,
    )
    mcts.run(progress_bar=True)
    print(mcts)

    evaluated_state = mcts.get_evaluated_state()
    evaluated_actions = mcts.get_evaluated_actions()

    evaluated_state_str = serialize_evaluated_state(evaluated_state=evaluated_state)
    evaluated_actions_str = serialize_evaluated_actions(evaluated_actions=evaluated_actions)

    evaluation_str = f"{evaluated_state_str} {evaluated_actions_str}"
    print(evaluation_str)

    with open(args.output_path, "w") as f:
        f.write(evaluation_str + "\n")

    print(f"evaluation saved to {args.output_path} successfully!")


if __name__ == "__main__":
    main()

File: utttcpp/src/mcts-evaluate.cpp
#include <cstdint>
#include <fstream>
#include <iostream>
#include <sstream>
#include <stdexcept>

#include "utttcpp/game/constants.hpp"
#include "utttcpp/game/ultimateTicTacToe.hpp"
#include "utttcpp/helpers/prngMersenneTwister.hpp"
#include "utttcpp/selfplay/monteCarloTreeSearch.hpp"

using utttcpp::STATE_SIZE;

using utttcpp::UltimateTicTacToe;

using utttcpp::InitializeMersenneTwisterPRNG;

using utttcpp::MonteCarloTreeSearch;
using utttcpp::EvaluatedState;
using utttcpp::serializeEvaluatedState;
using utttcpp::EvaluatedAction;
using utttcpp::serializeEvaluatedActions;

int parseInt(char *argv) {
    int number;
    std::stringstream stream(argv);
    stream >> number;
    return number;
}

double parseDouble(char *argv) {
    double number;
    std::stringstream stream(argv);
    stream >> number;
    return number;
}

std::string parseString(char *argv) {
    return std::string(argv);
}

uint32_t parseUInt32(char *argv) {
    uint32_t number;
    std::stringstream stream(argv);
    stream >> number;
    return number;
}

UltimateTicTacToe *parseUttt(char *argv) {
    unsigned char *state = new unsigned char[STATE_SIZE];
    std::string utttState = parseString(argv);
    int stateValue;
    for (int i = 0; i < STATE_SIZE; i++) {
        stateValue = utttState[i] - '0';
        if (stateValue < 0 || stateValue > 9) {
            throw std::runtime_error("invalid state value");
        }
        state[i] = (unsigned char)stateValue;
    }
    UltimateTicTacToe *uttt = new UltimateTicTacToe();
    uttt->setState(state);
    delete[] state;
    return uttt;
}

int main(int argc, char *argv[]) {
    if (argc != 6) {
        throw std::runtime_error(
            "required arguments: "
            "utttState, numSimulations, explorationStrength, prngSeed, outputPath");
    }

    std::cout << "utttState: " << argv[1] << std::endl;
    UltimateTicTacToe *uttt = parseUttt(argv[1]);
    std::cout << uttt->toString() << std::endl;

    const int numSimulations = parseInt(argv[2]);
    std::cout << "numSimulations: " << numSimulations << std::endl;

    const double explorationStrength = parseDouble(argv[3]);
    std::cout << "explorationStrength: " << explorationStrength << std::endl;

    const uint32_t prngSeed = parseUInt32(argv[4]);
    std::cout << "prngSeed: " << prngSeed << std::endl;
    InitializeMersenneTwisterPRNG(prngSeed);

    const std::string outputPath = parseString(argv[5]);
    std::cout << "outputPath: " << outputPath << std::endl;

    MonteCarloTreeSearch *mcts =
        new MonteCarloTreeSearch(uttt, numSimulations, explorationStrength);
    mcts->run();
    std::cout << mcts->toString() << std::endl;

    EvaluatedState *evaluatedState = mcts->getEvaluatedState();
    std::string evaluatedStateString = serializeEvaluatedState(evaluatedState);

    EvaluatedAction **evaluatedActions = mcts->getEvaluatedActions();
    int numEvaluatedActions = mcts->getNumEvaluatedActions();
    std::string evaluatedActionsString =
        serializeEvaluatedActions(evaluatedActions, numEvaluatedActions);

    std::string evaluationString = evaluatedStateString + " " + evaluatedActionsString;
    std::cout << evaluationString << std::endl;

    std::ofstream ofs(outputPath);
    ofs << evaluationString << std::endl;
    ofs.close();

    std::cout << "evaluation saved to " << outputPath << " successfully!" << std::endl;

    for (int i = 0; i < numEvaluatedActions; i++) {
        delete evaluatedActions[i];
    }
    delete[] evaluatedActions;
    delete[] evaluatedState->state;
    delete evaluatedState;
    delete mcts;

    return 0;
}
